dot-license-contracts/contracts/LicenseSale.sol

```diff
@@ -59,63 +85,121 @@ contract LicenseSale is LicenseOwnership {
     }
   }
 
+  function _performRenewal(uint256 _tokenId, uint256 _numCycles) internal {
+    // You cannot renew a non-expiring license
+    // ... but in what scenario can this happen?
+    // require(licenses[_tokenId].expirationTime != 0);
+    uint256 productId = licenseProductId(_tokenId);
+
+    // If our expiration is in the future, renewing adds time to that future expiration
+    // If our expiration has passed already, then we use `now` as the base.
+    uint256 renewalBaseTime = Math.max256(now, licenses[_tokenId].expirationTime);
+
+    // We assume that the payment has been validated outside of this function
+    uint256 newExpirationTime = renewalBaseTime.add(intervalOf(productId).mul(_numCycles));
+
+    licenses[_tokenId].expirationTime = newExpirationTime;
+
+    LicenseRenewal(
+      ownerOf(_tokenId),
+      _tokenId,
+      productId,
+      newExpirationTime
+    );
+  }
+
+  function _affiliateProgramIsActive() internal view returns (bool) {
+    return
+      affiliateProgram != address(0) &&
+      affiliateProgram.storeAddress() == address(this) &&
+      !affiliateProgram.paused();
+  }
+
   /** executives **/
   function setAffiliateProgramAddress(address _address) public onlyCEO {
     AffiliateProgram candidateContract = AffiliateProgram(_address);
     require(candidateContract.isAffiliateProgram());
     affiliateProgram = candidateContract;
   }
 
+  function setRenewalsCreditAffiliatesFor(uint256 _newTime) public onlyCEO {
+    renewalsCreditAffiliatesFor = _newTime;
+  }
+
   function createPromotionalPurchase(
     uint256 _productId,
+    uint256 _numCycles,
     address _assignee,
     uint256 _attributes
     )
     public
     onlyCOO
     whenNotPaused
-    returns (uint)
+    returns (uint256)
+  {
+    return _performPurchase(_productId, _numCycles, _assignee, _attributes, address(0));
+  }
+
+  function createPromotionalRenewal(
+    uint256 _tokenId,
+    uint256 _numCycles
+    )
+    public
+    onlyCOO
```

Comment:
Is the assumption that single users will assign themselves all roles to a single address? Otherwise restricting this to COO only makes it painful to manage. Is there a reason to avoid giving CEO everything?

Reply:
I suppose you're right -- the CEO can set any of the other positions, so limiting the CEO's power on any one method doesn't really provide any extra accountability-security.